### PR TITLE
Add placeholder types

### LINKED ISSUE #48696: handle universes in canonicalization
Now that https://github.com/rust-lang/rust/pull/47861/ has landed, there is a notion of universes for type variables. For now, in order to get #48411 landed, the canonicalization code is simply using `UniverseIndex::ROOT` for the universes of type variables it creates (those universes aren't really being used anyway presently).

However, what we *ought* to be doing is recording (during canonicalization) the universe of the variables that are being fed in. We can then canonicalize those universes (what chalk calls `u_canonicalize`) and then recreate them on the other side. And finally we map the universes back.

I'll try to leave more detailed notes later. I expect to either do this or mentor this as follow-up.

cc @sgrif

### LINKED ISSUE #55098: implement "instantiation" methods in chalk integration impl
There are a few methods in the "chalk integration" impl that have the job of "instantiating" canonical things with inference variables and the like. Actually, these methods have a somewhat bigger job: they are also supposed to create the surrounding inference context.

There are two such methods:

- [`instantiate_ucanonical_goal`]
- [`instantiate_ex_clause`]

[`instantiate_ucanonical_goal`]: https://github.com/rust-lang/rust/blob/e1643a8968753226dab7ab3c9fee826f097550f2/src/librustc_traits/chalk_context.rs#L149-L156
[`instantiate_ex_clause`]: https://github.com/rust-lang/rust/blob/e1643a8968753226dab7ab3c9fee826f097550f2/src/librustc_traits/chalk_context.rs#L157-L164

PR #54858 introduces a `enter_with_canonical` method that is more-or-less *exactly* what we need here ([link](https://github.com/rust-lang/rust/pull/54858/commits/fd046a2ede1f83383e7a4a5f653f75550db91ea4#diff-d6504a1a8d276f60cd1bffdf88be7be0R498)). Hopefully that will land soon. 

For reference, [the implementations of these methods from the chalk project are here.](https://github.com/rust-lang-nursery/chalk/blob/f615de34fb6e71fa8b969a6a804680533b2b9a85/chalk-solve/src/solve/slg/implementation.rs#L72-L93)

### PR DESCRIPTION
**Pull request by scalexm**
Fixes #48696 (handle universes in canonicalization of type inference vars), and fixes #55098.

### CODE REVIEW

**Review by nikomatsakis — APPROVED** [review]
beautiful

File: src/librustc/infer/canonical/canonicalizer.rs [hunk 1]
```diff
@@ -339,18 +339,49 @@ impl<'cx, 'gcx, 'tcx> TypeFolder<'gcx, 'tcx> for Canonicalizer<'cx, 'gcx, 'tcx>
 
     fn fold_ty(&mut self, t: Ty<'tcx>) -> Ty<'tcx> {
         match t.sty {
-            ty::Infer(ty::TyVar(_)) => self.canonicalize_ty_var(CanonicalTyVarKind::General, t),
+            ty::Infer(ty::TyVar(vid)) => {
+                match self.infcx.unwrap().probe_ty_var(vid) {
+                    // `t` could be a float / int variable: canonicalize that instead
```
[the remark this answers is not in the document]
Comment (scalexm) [reply src/librustc/infer/canonical/canonicalizer.rs]: Yes the goal was to extract the `ui`!

**Comment by scalexm** [comment]
I don't see the above failure locally, but it seems weird, maybe spurious. Anyway I rebased onto master just in case.
@bors r=nikomatsakis

**Comment by kennytm** [comment]
Same error, should be legit.

**Comment by scalexm** [comment]
Anyone here who might have already dealt with that test? It seems that the faulty line is somehow cut, e.g. in the debug output there are things like:
```
rustc::ty::fold: HasTypeFlagsVisitor: t=fn(fn(), isize, *const *const u8) -> isize {std::rt::lang_start::<()>} t.flags=(empty) self.flags=HAS_PROJECTION
```
but the fact that the travis log is skipping everything before that faulty line is not helping.

**Comment by scalexm** [comment]
Also now I remember than in a previous PR, I *had* this test failing locally but not on travis.

**Comment by scalexm** [comment]
Ok so I've just understood that the `<<<<<< SKIPPED 228531248 BYTES >>>>>>` does not come from travis but from `compiletest` instead:
https://github.com/rust-lang/rust/blob/4632cf2a2e8c1720e3ae06920d945c4ac8ecbbd3/src/tools/compiletest/src/runtest.rs#L3383

What happens is that some bytes are skipped in the output, and *very unfortunately* the output resumes at `{std::rt::lang_start::<()>}`, which is then interpreted as some JSON by `json::extract_rendered`:
https://github.com/rust-lang/rust/blob/4632cf2a2e8c1720e3ae06920d945c4ac8ecbbd3/src/tools/compiletest/src/json.rs#L79-L92

So I guess this is actually the same as #53332. Btw that above code seems very hacky and I think this should be fixed instead of trying to work around `read2_abbreviated`.
cc @kennytm, @alexcrichton, any thoughts on this?

**Comment by scalexm** [comment]
cc @kennytm see my above hack, which according to this code:
https://github.com/rust-lang/rust/blob/4632cf2a2e8c1720e3ae06920d945c4ac8ecbbd3/src/tools/compiletest/src/runtest.rs#L2690-L2709

should prevent `compiletest` from calling `json::extract_rendered(&proc_res.stderr, &proc_res)` for that particular test.

**Comment by kennytm** [comment]
Let's try.

@bors r=nikomatsakis

**Comment by nnethercote** [comment]
Instructions on how to run `keccak` and all the other benchmarks are [https://github.com/nnethercote/rustc-perf/tree/master/collector](here).

**Comment by scalexm** [comment]
Thanks @nnethercote for this detailed report.
I’m wondering whether #55517 may exhibit a similar perf regression.

**Comment by pnkfelix** [comment]
@nnethercote do you know offhand if PR #56251 addressed the performance regression that you identified above?

**Comment by nnethercote** [comment]
I compared 2018-11-25 (37961dbd) against 2018-11-25 (abe19a73), and there was a big regression for `keccak`, a small regression for `html5ever`, and the rest was basically noise.

The "this chart" you mentioned compares 2018-11-25 (37961dbd) against 2018-11-27 (aeff91d9), i.e. it's the same starting point, but a later end point. There is no regression, and a lot of improvements. So something has fixed the problem.

A more obvious way to see it is by consulting perf.rust-lang.org:
keccak
If you zoom in and look at the commit immediately after the drop, you can see that it was #56521 that fixed things. Nice job, @scalexm!

**Comment by eddyb** [comment]
@nnethercote #56521 doesn't exist, what PR did you mean?

**Pull request merged**